### PR TITLE
Add decimal128 support to Parquet reader and writer

### PR DESCRIPTION
**Pull request by vuule**
Closes https://github.com/rapidsai/cudf/issues/9566
Depends on https://github.com/rapidsai/cudf/pull/9804

Read decimal columns as 128bit when the input width requires it.
Write decimal128 columns as `FIXED_LEN_BYTE_ARRAY`.
Use the smallest viable decimal size to read `FIXED_LEN_BYTE_ARRAY` (used to default to decimal64, even when 32bits are sufficient).
Removes `strict_decimal_types` option from Parquet reader, we can now always read using the exact decimal type.

### CODE REVIEW

**Comment by vuule** [comment]
rerun tests

**Review by galipremsagar — APPROVED** [review]

File: cpp/tests/io/parquet_test.cpp [hunk 1]
```diff
@@ -3016,7 +3021,7 @@ TEST_F(ParquetReaderTest, DecimalRead)
     cudf::test::expect_columns_equal(result.tbl->view().column(1), col1);
 
     read_opts.set_columns({"dec20p1"});
-    EXPECT_THROW(cudf_io::read_parquet(read_opts), cudf::logic_error);
+    EXPECT_NO_THROW(cudf_io::read_parquet(read_opts));
```
Comment (devavret) [inline cpp/tests/io/parquet_test.cpp:3024]: If this doesn't throw then it should probably check if the results match

Comment (vuule) [reply cpp/tests/io/parquet_test.cpp:3024]: This test uses an embedded file, so I don't have the reference output available. Any suggestions?

Comment (devavret) [reply cpp/tests/io/parquet_test.cpp:3024]: 1. Write out the embedded file with host_write
2. Read it using pandas and grab the third column
3. ...
4. Profit!?

File: java/src/main/native/src/TableJni.cpp [hunk 2]
```diff
@@ -1299,9 +1299,11 @@ JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_Table_readCSV(
...
 }
 
-JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_Table_readParquet(
-    JNIEnv *env, jclass, jobjectArray filter_col_names, jstring inputfilepath, jlong buffer,
-    jlong buffer_length, jint unit, jboolean strict_decimal_types) {
+JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_Table_readParquet(JNIEnv *env, jclass,
+                                                                   jobjectArray filter_col_names,
+                                                                   jstring inputfilepath,
+                                                                   jlong buffer,
+                                                                   jlong buffer_length, jint unit) {
```
Comment (mythrocks) [inline java/src/main/native/src/TableJni.cpp:1306]: This is a breaking change in signature. I am surprised that the Java tests are running. I would have expected a corresponding change in the `Table` Java class's native method signature.

Comment (kuhushukla) [reply java/src/main/native/src/TableJni.cpp:1306]: Good point. When looking at the java side maybe I can take this into consideration. I need to evaluate how much of a change it is on java and plugin side.

Comment (mythrocks) [reply java/src/main/native/src/TableJni.cpp:1306]: [Here](https://github.com/rapidsai/cudf/blob/branch-22.02/java/src/main/java/ai/rapids/cudf/Table.java#L247) is the declaration of the native method in the Java `Table` class.

[Here](https://github.com/rapidsai/cudf/blob/branch-22.02/java/src/main/java/ai/rapids/cudf/Table.java#L868) is where that native call is invoked.

The fix might be as simple as to drop the last parameter in the native function declaration, and remove that argument at the invocation site.

Comment (kuhushukla) [reply java/src/main/native/src/TableJni.cpp:1306]: Absolutely. Just wanted to be sure of invocation sites but rest looks straight forward.

Comment (vuule) [reply java/src/main/native/src/TableJni.cpp:1306]: Thanks for the pointers. I'll update the call sites. 👍

**Review by devavret — APPROVED** [review]

File: java/src/main/native/src/TableJni.cpp (hunk 2 above)
Comment (vuule) [reply java/src/main/native/src/TableJni.cpp:1306]: Made the updates. The parameter was used in only one test.

**Review by mythrocks — APPROVED** [review]
Thank you for accommodating the changes. LGTM!

File: cpp/tests/io/parquet_test.cpp [hunk 3]
```diff
@@ -3020,8 +3018,34 @@ TEST_F(ParquetReaderTest, DecimalRead)
       std::begin(col1_data), std::end(col1_data), validity_c1, numeric::scale_type{-11});
     cudf::test::expect_columns_equal(result.tbl->view().column(1), col1);
 
-    read_opts.set_columns({"dec20p1"});
-    EXPECT_NO_THROW(cudf_io::read_parquet(read_opts));
+    auto validity_c2 =
+      cudf::detail::make_counting_transform_iterator(0, [](auto i) { return i != 6 and i != 14; });
```
Comment (mythrocks) [inline cpp/tests/io/parquet_test.cpp:3022]: Apologies for missing this on the first pass. 
We might consider rephrasing this:
```c++
auto validity_c2 = cudf::test::iterators::nulls_at({6, 14});
```

File: java/src/main/native/src/TableJni.cpp (hunk 2 above)
Comment (mythrocks) [reply java/src/main/native/src/TableJni.cpp:1306]: > Absolutely. Just wanted to be sure of invocation sites but rest looks straight forward.

:] @kuhushukla: Sorry, you might have found it strange that I was pointing out the Java code to you (of all people). That comment was addressed to @vuule. The confusion is regretted. :D

File: cpp/src/io/parquet/page_enc.cu [hunk 4]
```diff
@@ -72,6 +72,22 @@ struct page_enc_state_s {
   uint16_t vals[rle_buffer_size];
 };
 
+/**
+ * @brief Returns the size of the type in the Parquet file.
+ */
+uint32_t __device__ physical_type_len(Type physical_type, type_id id)
+{
+  if (physical_type == FIXED_LEN_BYTE_ARRAY and id == type_id::DECIMAL128)
+    return sizeof(__int128_t);
```
Comment (nvdbaranec) [inline cpp/src/io/parquet/page_enc.cu:81]:
```suggestion
  if (physical_type == FIXED_LEN_BYTE_ARRAY and id == type_id::DECIMAL128){
    return sizeof(__int128_t);
  }
```

**Review by nvdbaranec — COMMENTED** [review]
Looks great.  I felt the need to add one change, just to be annoying.

**Review by charlesbluca — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

**Review by nvdbaranec — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**